### PR TITLE
fix(protocols): run compaction even under highload

### PR DESCRIPTION
**Pull request by ChrisKujawa**
* compaction was skipped under highload even if we are going out of memory - because DISK storage level was not considered and wrong values for memory are used

See related issue https://github.com/zeebe-io/zeebe/issues/3090

### CODE REVIEW

File: protocols/raft/src/main/java/io/atomix/protocols/raft/impl/RaftContext.java [hunk 1]
```diff
@@ -154,7 +170,10 @@ public RaftContext(
     this.threadContextFactory = checkNotNull(threadContextFactory, "threadContextFactory cannot be null");
     this.closeOnStop = closeOnStop;
 
-    this.loadMonitor = new LoadMonitor(LOAD_WINDOW_SIZE, HIGH_LOAD_THRESHOLD, loadContext);
+    this.loadMonitor =
```
Comment (npepinpe) [inline protocols/raft/src/main/java/io/atomix/protocols/raft/impl/RaftContext.java]: I'd rather we do like the other things and add a `checkNotNull`, ensure it's built already before (so like the `RaftStateMachineFactory` we just add a default value which can be `LoadMonitor::new`?) Makes more sense to me to not pass null arguments

File: protocols/raft/src/main/java/io/atomix/protocols/raft/storage/RaftStorage.java [hunk 2]
```diff
@@ -105,7 +106,7 @@ private RaftStorage(
     this.freeMemoryBuffer = freeMemoryBuffer;
     this.flushOnCommit = flushOnCommit;
     this.retainStaleSnapshots = retainStaleSnapshots;
-    this.statistics = new StorageStatistics(directory);
+    this.statistics = storageStatistics == null ? new StorageStatistics(directory) : storageStatistics;
```
Comment (npepinpe) [inline protocols/raft/src/main/java/io/atomix/protocols/raft/storage/RaftStorage.java]: Can we move this logic to the builder?

**Comment by npepinpe** [comment]
Very nice

I tested it and now the memory is rock solid! :fire: :fire: 

Screenshot from 2019-09-26 16-26-14

This is taken over ~1 hour, and you can see the heap is fluctuating between ~500mb to ~2gb, and once load is taken off it goes back to ~500mb. I also didn't see a single failover under high load, so that issue might be related to this (i.e. not compacting enough maybe, or using too much memory putting pressure on the GC?).

**Review by npepinpe — APPROVED** [review]
:tada: 

One note (not change request), seems the load monitor is really only used in the service manager, even though it's created in the context (for some reason). Might be simpler to move it to the service manager, which lets you unit test that instead of the more complicated Raft test, then we don't need a factory for it.

File: protocols/raft/src/main/java/io/atomix/protocols/raft/storage/RaftStorage.java (hunk 2 above)
Comment (ChrisKujawa) [reply protocols/raft/src/main/java/io/atomix/protocols/raft/storage/RaftStorage.java]: Yes see latest commit

File: protocols/raft/src/main/java/io/atomix/protocols/raft/impl/RaftContext.java (hunk 1 above)
Comment (ChrisKujawa) [reply protocols/raft/src/main/java/io/atomix/protocols/raft/impl/RaftContext.java]: I get what you mean I was not sure If I want to always create an object even if there will be one which is replacing this. Btw there is not checkNotNull for the stateMachineFactory

**Comment by ChrisKujawa** [comment]
> One note (not change request), seems the load monitor is really only used in the service manager, even though it's created in the context (for some reason). Might be simpler to move it to the service manager, which lets you unit test that instead of the more complicated Raft test, then we don't need a factory for it.

Totally agree, but currently the `RaftContext` is used in the way it keeps all the stuff the other services/classes need otherwise we will blow up our ctor arguments etc. I would like to keep it for now like this, but yes it is not ideal. I also would like to keep the raft test since now it tests e2e the problematic points. If I move this in the Manager I'm not sure If I can still test this in that way.

**Comment by ChrisKujawa** [comment]
I spent the last hour to write an test for the RaftServiceManager it was pain in the ass because everything is coupled and can't be replace from outside (most of the things we need to change for the tests is created inside the context or manager) like SNAPSHOT interval or SingleThreadContext etc. I can show you the last thing I came up with but it is not done yet. I think we should improve/refactor here so we can test this better.

File: protocols/raft/src/main/java/io/atomix/protocols/raft/impl/RaftContext.java (hunk 1 above)
Comment (ChrisKujawa) [reply protocols/raft/src/main/java/io/atomix/protocols/raft/impl/RaftContext.java]: I changed it

**Review by npepinpe — APPROVED** [review]
:+1: 

The code generally looks fine but now a lot of tests failed with OutOfMemory errors, though I don't see anything special from the changes you added now.

**Comment by ChrisKujawa** [comment]
> The code generally looks fine but now a lot of tests failed with OutOfMemory errors, though I don't see anything special from the changes you added now.

Hmm was there some limit changes on our jenkins? I saw lately also out of mem on Zeebe builds

**Pull request merged**